File: src/test/java/com/dunamismax/site/content/ContentServiceTest.java
package com.dunamismax.site.content;

import static org.assertj.core.api.Assertions.assertThat;

import java.util.Comparator;
import org.junit.jupiter.api.Test;

class ContentServiceTest {

  private final ContentService service = service();

  @Test
  void projectsLoadWithAtLeastOneEntry() {
    assertThat(service.projects()).isNotEmpty();
  }

  @Test
  void projectsAreSortedByCategoryPositionThenName() {
    service.projectsByCategory().values().forEach(group -> {
      var sorted = group.stream()
          .sorted(Comparator.comparingInt(Project::position).thenComparing(Project::name))
          .toList();
      assertThat(group).containsExactlyElementsOf(sorted);
    });
  }

  @Test
  void featuredProjectsAreASubsetOfAllProjects() {
    var all = service.projects();
    assertThat(all).containsAll(service.featuredProjects());
    assertThat(service.featuredProjects()).allMatch(Project::featured);
  }

  @Test
  void aboutHtmlIsRenderedFromMarkdown() {
    String html = service.aboutHtml();
    assertThat(html).contains("<strong>Go</strong>");
    assertThat(html).contains("<strong>C</strong>");
    assertThat(html).contains("PostgreSQL");
  }

  private static ContentService service() {
    var service = new ContentService(new MarkdownRenderer());
    service.load();
    return service;
  }
}

File: src/test/kotlin/com/dunamismax/site/content/ContentServiceTest.kt
package com.dunamismax.site.content

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

class ContentServiceTest {

    private val service = ContentService(MarkdownRenderer()).also { it.load() }

    @Test
    fun `projects load with at least one entry`() {
        assertThat(service.projects()).isNotEmpty
    }

    @Test
    fun `projects are sorted by category position then name`() {
        val byCategory = service.projectsByCategory()
        byCategory.values.forEach { group ->
            val sorted = group.sortedWith(compareBy({ it.position }, { it.name }))
            assertThat(group).containsExactlyElementsOf(sorted)
        }
    }

    @Test
    fun `featured projects are a subset of all projects`() {
        val all = service.projects().toSet()
        assertThat(all).containsAll(service.featuredProjects())
        assertThat(service.featuredProjects()).allMatch { it.featured }
    }

    @Test
    fun `about html is rendered from markdown`() {
        val html = service.aboutHtml()
        assertThat(html).contains("Java 25 LTS")
        assertThat(html).contains("PostgreSQL")
    }
}
